src/db/model/sync.rs

```diff
@@ -17,37 +19,90 @@ use crate::db::MongoDb;
 }
 
 impl SyncRecord {
     /// The status collection name.
     pub const COLLECTION: &'static str = "sync";
 }
 
+/// An aggregation type that represents the ranges of completed milestones and gaps.
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+pub struct SyncData {
+    /// The completed(synced and logged) milestones data
+    pub completed: Vec<Range<u32>>,
+    /// Gaps/missings milestones data
+    pub gaps: Vec<Range<u32>>,
+}
+
 impl MongoDb {
+    /// If available, returns the [`SyncRecord`] associated with the provided milestone `index`.
+    pub async fn get_sync_record_by_index(&self, index: u32) -> Result<Option<SyncRecord>, Error> {
+        self.0
+            .collection::<SyncRecord>(SyncRecord::COLLECTION)
+            .find_one(doc! {"milestone_index": index}, None)
+            .await
+    }
+
     /// Upserts a [`SyncRecord`] to the database.
-    pub async fn upsert_sync_record(&self, record: &SyncRecord) -> Result<UpdateResult, Error> {
+    pub async fn upsert_sync_record(&self, index: u32) -> Result<UpdateResult, Error> {
         self.0
             .collection::<SyncRecord>(SyncRecord::COLLECTION)
             .update_one(
-                doc! {"milestone_index": record.milestone_index},
-                doc! {"$set": bson::to_document(record)?},
+                doc! {"_id": index},
+                doc! {"$set": bson::to_document(&SyncRecord{milestone_index: index})?},
                 UpdateOptions::builder().upsert(true).build(),
             )
             .await
     }
     /// Retrieves the sync records sorted by [`milestone_index`](SyncRecord::milestone_index).
-    pub async fn sync_records_sorted(&self) -> Result<impl Stream<Item = Result<SyncRecord, Error>>, Error> {
+    pub async fn sync_records_sorted(
+        &self,
+        start: u32,
+        end: u32,
+    ) -> Result<impl Stream<Item = Result<SyncRecord, Error>>, Error> {
         self.0
             .collection::<SyncRecord>(SyncRecord::COLLECTION)
             .find(
-                doc! { "synced": true },
-                FindOptions::builder().sort(doc! {"milestone_index": 1u32}).build(),
+                doc! { "milestone_index": { "$gte": start, "$lte": end } },
+                FindOptions::builder().sort(doc! {"milestone_index": 1}).build(),
             )
             .await
     }
+
+    /// Retrieves a [`SyncData`] structure that contains the completed and gaps ranges.
+    pub async fn get_sync_data(&self, start: u32, end: u32) -> Result<SyncData, Error> {
+        let mut res = self.sync_records_sorted(start, end).await?;
+        let mut sync_data = SyncData::default();
+        let mut last_record: Option<u32> = None;
+        while let Some(SyncRecord { milestone_index }) = res.try_next().await? {
+            // Missing records go into gaps
+            if let Some(last) = last_record.as_ref() {
+                if last + 1 != milestone_index {
+                    sync_data.gaps.push(last + 1..milestone_index - 1);
```

Comment:
Shouldn't these be inclusive ranges throughout?

Reply:
They are inclusive. This gap is found when a record exists, therefore it is not part of the gap.

Reply:
Hmm, I'm confused. Aren't you using a `Range` (exclusive) to represent an inclusive range that should be represented using `RangeInclusive`?

Reply:
Well, yes, but that property of ranges only applies when they are used as an iterator. I am using `range.start` and `range.end`. However, we can change it for clarity.